Домашнее задание №4, "Exceptions Logging". Первая попытка сдачи.

IgushkinV opened the pull request:
Переделал. Теперь файлы из пакета Lesson2, с изменениями.

IgushkinV committed:
Домашнее задание №4, "Exceptions Logging". Первая попытка сдачи.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Main.java:
```diff
@@ -11,20 +14,61 @@ public static void main(String[] args) {
-        stringStorage.delete();
-        System.out.println(stringStorage.getLast());
+        try {
+            System.out.println("Элемент хранилища c индексом 3: " + stringStorage.get(3));
```
у нас ведь уже есть логгеры, может перейдем на них?

IgushkinV replied:
Я так понимаю, что здесь нужно создать консольный аппендер для класса Main. Весь вывод с помощью System.out.println заменить?

tlt-java-courses replied:
Да, весь sout лучше поменять. Отдельный апендер, не обязательно. Просто мы уже умеем в логеры, так что нужно уходить от консоли.

IgushkinV replied:
Сделано.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Main.java:
```diff
@@ -11,20 +14,61 @@ public static void main(String[] args) {
+        try {
+            oneElemStorage.getLast();
+        } catch (NegativeIndexException e) {
+            e.printStackTrace();
```
Может что-то напишем в лог или закинем туда наш эксепшн?

IgushkinV replied:
Я при генерации каждого исключения передаю в конструктор String, в котором объясняется, какая ошибка произошла. И там же, строкой выше генерации исключения, пишу в лог сообщение об ошибке (почти дублирую Stirng из исключения). Это неправильная практика?

tlt-java-courses replied:
Если ты хочешь что-то распечатать, а в данном случае эксепшн, то лучше передавать его в логгер, он распечатает и сообщение которое ты передал в конструкторе и стек трейс вместе с ним.

IgushkinV replied:
В лог так же записывать сообщение при создании исключения (в месте выброса из метода)? Или оставить только записать при обработке исключения?

tlt-java-courses replied:
при обоработке

IgushkinV replied:
Что-то такая каша получается, если выводить в консоль логи и из места создания исключения, и из места обработки что-то писать, и еще при передаче исключения в лог опять текст выводится...

IgushkinV replied:
Сделано.

tlt-java-courses replied:
В месте создания исключения тебе не нужно печатать об этом, если ты имеешь в виду, что ты передаешь строку в конструктор, то там описывается краткое описание ошибки и наш эксепшн хранит ее в себе, печать данных в лог должна быть в месте обработки этого исключения.
В данном случае у тебя это происходит в блоке catch, т.е его обработке.

IgushkinV replied:
Понял, сделал.

tlt-java-courses commented on src/main/java/Igushkin/Lesson2/Main.java:
```diff
@@ -11,20 +14,61 @@ public static void main(String[] args) {
+            nameStorage.get(-1);
+        } catch (NegativeIndexException e) {
+            e.printStackTrace();
+            log.warn("Запрос элемента с отрицательным индексом");
```
Если мы закинем ошибку в лог, то она напечатает стек трейс, давай попробуем?

IgushkinV replied:
Сделано.

IgushkinV committed:
Исправление замечаний к ДЗ №4.
Заменил sout на лог.
Добавил exception в лог для вывода стектрейса в логе.
Добавил вывод сообщения при обработке исключений.

IgushkinV committed:
Убрал запись логов в местах генерации исключений в классе Storage.
